### ISSUE TITLE
Proposal: C422 Unnecessary filter() usage

### ISSUE THREAD

**Issue by grihabor** [astral-sh/ruff#18492]
## What it does

Checks for unnecessary `filter()` calls with lambda functions.

Why is this bad?

Using `filter(func, iterable)` when func is a lambda is slower than using a generator expression or a comprehension, as the latter approach avoids the function call overhead, in addition to being more readable.

This rule also applies to `filter()` calls within `list()`, `set()`, and `dict()` calls. For example:

- Instead of `list(filter(lambda num: num < 10, nums))`, use `[num for num in nums if num < 10]`.
- Instead of `set(filter(lambda num: num % 2 == 0, nums))`, use `{num for num in nums if num % 2 == 0}`.
- Instead of `dict(filter(lambda pair: pair[1] is not None, items))`, use `{k: v for k, v in values if v is not None}`.

## Example
```python
filter(lambda x: x % 2 == 0, iterable)
```
Use instead:
```python
(x for x in iterable if x % 2 == 0)
```

**Comment by ntBre** [comment]
Is this rule present in the upstream flake8-comprehensions? I checked the [PyPI page](https://pypi.org/project/flake8-comprehensions/), and it looked like it stopped at C420. We could still consider the rule, but we'd want to be careful adding it as a `C` rule without an upstream feature request.

**Comment by grihabor** [comment]
No, it's not in upstream, but it fits the overall theme of C4 checks. Should I open an proposal in upstream? Or should we choose a different rule tag?

**Comment by ntBre** [comment]
I think it may be okay in light of [this comment](https://github.com/astral-sh/ruff/issues/18331#issuecomment-2922513633) from the flake8-comprehensions maintainer last week. I forgot that was for the same linter.

**Comment by grihabor** [comment]
Ok, so C421 is taken, let's reserve C422 then.

**Comment by grihabor** [comment]
Hey @adamchainz, do you have any thoughts on this rule?

**Comment by SpecLad** [comment]
FWIW, I had filed an upstream feature request for this: adamchainz/flake8-comprehensions#622.
